app/auth/views/proton.py

```diff
@@ -0,0 +1,119 @@
+import json
+
+from flask import request, session, redirect, flash, url_for
+from flask_login import current_user
+from requests_oauthlib import OAuth2Session
+from typing import Optional
+
+from app.auth.base import auth_bp
+from app.auth.views.login_utils import after_login
+from app.config import (
+    PROTON_BASE_URL,
+    PROTON_CLIENT_ID,
+    PROTON_CLIENT_SECRET,
+    PROTON_VALIDATE_CERTS,
+    URL,
+)
+from app.proton.proton_client import HttpProtonClient, convert_access_token
+from app.proton.proton_callback_handler import ProtonCallbackHandler, Action
+from app.utils import encode_url
+
+_authorization_base_url = PROTON_BASE_URL + "/oauth/authorize"
+_token_url = PROTON_BASE_URL + "/oauth/token"
+
+# need to set explicitly redirect_uri instead of leaving the lib to pre-fill redirect_uri
+# when served behind nginx, the redirect_uri is localhost... and not the real url
+_redirect_uri = URL + "/auth/proton/callback"
+
+
+def get_client_ip() -> Optional[str]:
+    if request.environ.get("HTTP_X_FORWARDED_FOR") is None:
+        return request.environ.get("REMOTE_ADDR")
+    else:
+        return request.environ["HTTP_X_FORWARDED_FOR"]  # if behind a proxy
+
+
+def extract_action() -> Action:
+    action = request.args.get("action")
+    if action is not None:
+        if action == "link":
+            return Action.Link
+        else:
+            raise Exception(f"Unknown action: {action}")
+    return Action.Login
+
+
+def get_action_from_state() -> Action:
+    oauth_flow_state = json.loads(session["oauth_flow_state"])
+    oauth_action = oauth_flow_state.get("action")
+    if oauth_action == Action.Login.value:
+        return Action.Login
+    elif oauth_action == Action.Link.value:
+        return Action.Link
+    raise Exception(f"Unknown action in state: {oauth_action}")
+
+
+@auth_bp.route("/proton/login")
+def proton_login():
+    if PROTON_CLIENT_ID is None or PROTON_CLIENT_SECRET is None:
+        return redirect(url_for("auth.login"))
+
+    next_url = request.args.get("next")
```

Comment:
To prevent redirecting outside SL:

```suggestion
    next_url = sanitize_next_url(request.args.get("next"))
```

Reply:
Changed it, nice catch